### PR TITLE
avoid to banish region leaders when rolling upgrade

### LINKED ISSUE #12166: region leaders can be banished in an election timeout when rolling upgrade a cluster
### What version of TiKV are you using?
master

The log files is in the attachment. It shows a region 27073. At `18:28:43` one replica starts an pre-election, and then the the region leader is banished about 10s.

After some investigations, I think it's caused by an issue of hibernate regions.

[27073.zip](https://github.com/tikv/tikv/files/8252692/27073.zip)

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

### What is changed and how it works?

Issue Number: Close #12166 

Currently it only contains a test case. More commits will be pushed to fix the issue later.

What's Changed:

```commit-message
avoit to banish region leaders when rolling upgrade
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

### Release note 

```release-note
None
```

### CODE REVIEW

File: tests/failpoints/cases/test_hibernate.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    let mut cluster = new_node_cluster(0, 3);
+    let base_tick_ms = 50;
+    cluster.cfg.raft_store.raft_base_tick_interval = ReadableDuration::millis(base_tick_ms);
+    cluster.cfg.raft_store.raft_heartbeat_ticks = 2;
+    cluster.cfg.raft_store.raft_election_timeout_ticks = 10;
+    // So the random election timeout will always be 10, which makes the case more stable.
+    cluster.cfg.raft_store.raft_min_election_timeout_ticks = 10;
+    cluster.cfg.raft_store.raft_max_election_timeout_ticks = 11;
+    // Disable peer stale state check.
+    cluster.cfg.raft_store.abnormal_leader_missing_duration = ReadableDuration::secs(1000);
```
Comment (BusyJay) [inline tests/failpoints/cases/test_hibernate.rs:26]: You mean `configure_for_hibernate`?

File: tests/failpoints/cases/test_hibernate.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    let (tx, rx) = mpsc::sync_channel(128);
+    let filter = RegionPacketFilter::new(1, 3)
+        .direction(Direction::Send)
+        .msg_type(MessageType::MsgRequestVote)
+        .when(Arc::new(AtomicBool::new(false)))
+        .set_msg_callback(Arc::new(move |m| drop(tx.send(m.clone()))));
+    cluster.add_send_filter(CloneFilterFactory(filter));
+    cluster.run_node(2).unwrap();
+
+    // If peer 3 starts a new election, it can be rejected by peer 1 and 2 because both of them
```
Comment (BusyJay) [inline tests/failpoints/cases/test_hibernate.rs:82]: The comment is not valid anymore.

**Review by BusyJay — APPROVED** [review]

**Review by NingLin-P — APPROVED** [review]

**Pull request merged**